Mixed detector

rhiannonlynne opened the pull request:
Fixed m2throughput.dat file and ipython notebook to generate this file.

Added SEDs. 

Added sedUtils and bandpassUtils - note bandpassUtils will let you build throughput curves from base components! (ie.. they duplicate some of the functionality of the ipython notebooks and we should probably think about what to do with that). 

Added mixedDetector analysis (uses bandpassUtils and sedUtils, requires sims_photUtils).

rhiannonlynne committed:
Fixing problem with m2Throughput.dat file and updated notebook

rhiannonlynne committed:
Added utilities to combine base throughput + loss files and deal with seds in Seds directory.

rhiannonlynne committed:
Added mixedDetector photometry test script.

rhiannonlynne committed:
just reran detector file (no changes).

rhiannonlynne committed:
added ipynb_checkpoints to gitignore

rhiannonlynne committed:
Removed iPython checkpoints from repo.

rhiannonlynne commented:
I'd like to have @cclaver review this, and also Scott Daniels (@danielsf)

danielsf commented on python/bandpassUtils.py:
```diff
@@ -0,0 +1,251 @@
+    # Set SYSENG_THROUGHPUTS_DIR to the root dir for syseng_throughputs ('setup syseng_throughputs' will do this automatically)
+    defaultDirs = {}
+    if rootDir is None:
+        rootDir = os.getenv('SYSENG_THROUGHPUTS_DIR')
```
Do you want to support people who have not use eups to setup the stack?  If not, should this be

rootDir = lsst.utils.getPackageDir('sysengh_throughputs')

?

rhiannonlynne replied:
Yes, I will swap to this.

danielsf commented on python/bandpassUtils.py:
```diff
@@ -0,0 +1,251 @@
+
+def buildVendorDetector(vendorDir, addLosses=True):
+    """
+    Assumes there is a file *QE.dat and *Losses.dat to represent the QE response and the Losses.
```
This docstring should be more explicit, i.e. "if addLosses==True, multiplies losses and QE together in restuls; if addLosses==False, only returns QE"

rhiannonlynne replied:
Yeah - needs more documentation in general. Will add.

danielsf commented on python/bandpassUtils.py:
```diff
@@ -0,0 +1,251 @@
+
+def buildGenericDetector(detectorDir, addLosses=True):
+    """
+    Combine detector throughputs from multiple vendors, using the minimum value of QE*losses at each wavelength.
```
mention behavior of addLosses in docstring

danielsf commented on python/bandpassUtils.py:
```diff
@@ -0,0 +1,251 @@
+    qe.setBandpass(wavelen, sbMin)
+    return qe
+
+def buildFilters(filterDir, addLosses=True):
```
mention behavior of addLosses in docstring

danielsf commented on python/bandpassUtils.py:
```diff
@@ -0,0 +1,251 @@
+            filters[f].setBandpass(wavelen, sb)
+    return filters
+
+def savitzky_golay(y, window_size=31, order=3, deriv=0, rate=1):
```
Should this method have a docstring?

danielsf replied:
I it is not clear to me what deriv and rate are

rhiannonlynne replied:
This whole function was lifted from Chuck's ipython notebook. I actually don't know what it's supposed to do, other than smooth the wavelength response of the borosilicate glass throughput curve (and I don't know why you have to do that). I have a request to Chuck to clarify and to explain, but otherwise this whole function is just here because it looked like I needed it (see makeLens*.ipynb). Perhaps @cclaver can fill in some more information on this PR?

danielsf commented on python/bandpassUtils.py:
```diff
@@ -0,0 +1,251 @@
+    y = np.concatenate((firstvals, y, lastvals))
+    return np.convolve(m[::-1], y, mode='valid')
+
+def buildLens(lensDir, addLosses=True):
```
mention behavior of addLosses in docstring

danielsf replied:
Does the loss file sub-directory need to obey a specific naming convention?

danielsf commented on python/bandpassUtils.py:
```diff
@@ -0,0 +1,251 @@
+    lens.setBandpass(wavelen, sb)
+    # Add losses.
+    if addLosses:
+        lossfiles = glob(os.path.join(lensDir, 'l*Losses", ''*.dat'))
```
Does this line work?  I think the double- and single- quotation marks got mixed up.

rhiannonlynne replied:
It did somehow (results of the bandpass look like the version calculated by Chuck, so must include losses), but I will fix.

danielsf commented on python/bandpassUtils.py:
```diff
@@ -0,0 +1,251 @@
+        lens.setBandpass(wavelen, sb)
+    return lens
+
+def buildMirror(mirrorDir, addLosses=True):
```
mention addLosses in docstring

danielsf commented on python/bandpassUtils.py:
```diff
@@ -0,0 +1,251 @@
+    atmo.readThroughput(atmofile)
+    return atmo
+
+def buildHardwareAndSystem(defaultDirs, addLosses=True):
```
docstring should specify what defaultDirs is (a dict pointing to the directories holding the data files associated with each throughput component)

danielsf commented on python/bandpassUtils.py:
```diff
@@ -0,0 +1,251 @@
+        system[f].setBandpass(wavelen, hw_sb*atmosphere.sb)
+    return hardware, system
+
+def plotBandpasses(bandpassDict, title=None, newfig=True, savefig=False, addlegend=True,
```
Define arguments and kwargs in docstring

rhiannonlynne committed:
Responding to Scott's comments-  mostly doc updates

rhiannonlynne committed:
Reduced redshift range for photoz outliers

danielsf commented on python/sedUtils.py:
```diff
@@ -0,0 +1,273 @@
+    # The environment variable PHOT_REF_SEDS_DIR is set if the file
+    #  'phot_ref.csh' is sourced.
+    if sedDir == None:
+        sedDir = os.path.join(os.getenv('SYSENG_THROUGHPUTS_DIR'), 'seds')
```
Here's another instance where we should probably use lsst.utils.getPackageDir

danielsf commented on python/sedUtils.py:
```diff
@@ -0,0 +1,273 @@
+def redshiftSingleSED(sed_in, z):
+    # Make a copy of the input SED, as otherwise we will overwrite original arrays.
+    sed_out = deepcopy(sed_in)
+    sed_out.redshiftSED(z, dimming=False)
```
what is the rationale behind setting dimming=False?  In the catalogs, we setting dimming=True

danielsf commented on python/sedUtils.py:
```diff
@@ -0,0 +1,273 @@
+    return sedDict, sedlists
+
+
+def makeRedshiftedSeds(sedDict, sedlists, redshifts):
```
it looks like this method returns two dicts: one (sedDict) that is keyed on the SEDs individual name.  One (sedlists) that is keyed on the object type and whose values are a list of all the SEDs of that object type.  sedlists should be renamed so it is obvious that it is a dict.

danielsf replied:
Okay: sedlists is a dict of lists of sed names.  Is it possible to just work with discrete sedDicts (one for quasars, one for galaxies, one for stars).  Or even just make sedDict a dict of dicts, so that

sedDict ['quasar']['quasar_name'] 

was a quasar SED?  That way, you wouldn't have to keep carrying around both sedDict and sedlists

rhiannonlynne replied:
I think this makes sense. I had some other reason for doing this originally (a long time ago, when I was using these methods for the first time, but I maybe wasn't so smart at python then and anyway I don't remember why I did it in the first place .. but now is a good time to clean this up).

danielsf commented on python/sedUtils.py:
```diff
@@ -0,0 +1,273 @@
+    return sed_out
+
+
+def matchSedsBp(sedDict, bpDict, refFilter=None):
```
Should this method make use of Sed.setupPhiArray to further speed up the calculation?

danielsf replied:
There is now a class in sims_photUtils called BandpassDict which reads in a bunch of bandpasses, and then automatically resamples them to the same wavelength grid and calculates their phi array.  It uses Sed.manyMagCalc to calculate the magnitude of any Sed in all of the bandpasses simultaneously.  Maybe that is worth using here?

danielsf commented on python/sedUtils.py:
```diff
@@ -0,0 +1,273 @@
+    return sedDict, bpDict
+
+
+def calcNatMags(bandpassDict, sedDict, sedlists):
```
This method does not appear to use sedDict

danielsf replied:
wait.  Nevermind.  I see it.  I am very confused what the difference between sedDict and sedLists is.

danielsf commented on python/mixedDetector_photometry.py:
```diff
@@ -0,0 +1,171 @@
+    photParams = PhotometricParameters()
+    lsstDefaults = LSSTdefaults()
+    darksky = Sed()
+    darksky.readSED_flambda(os.path.join('../siteProperties', 'darksky.dat'))
```
This should also use lsst.utils.getPackageDir so that it does not depend on where it is being run from

danielsf commented:
I am effectively done with my review.

rhiannonlynne committed:
Added plot (with dark sky mags) to calcM5 section

rhiannonlynne committed:
Removed sedlists, changed to nested dict of sedDict

rhiannonlynne commented:
@cclaver any comments? I'd  like to get this merged to master.

rhiannonlynne committed:
Added effective wavelength calculation at many atmospheres

rhiannonlynne committed:
Added fit for effective wavelength with X

rhiannonlynne committed:
Renaming and reorganizing files.

rhiannonlynne committed:
Renaming and reorganizing files.

rhiannonlynne committed:
Moved notebooks to notebooks directory.

rhiannonlynne committed:
Update gitignore to properly ignore notebook checkpoint files.

rhiannonlynne committed:
Update plots generated by new scripts or notebooks.

rhiannonlynne committed:
Added new python utilities.

bandpassUtils deals with all components and combines them into bandpass objects.
calcM5 calculates all quantities needed for table2.
effective wavelength calculates the effective wavelength (at a range of airmasses)
mixedDetector_Photometry investigates a mixed detector focal plane.

rhiannonlynne commented:
Talked to @cclaver via email. Incorporated changes suggested by @danielsf and added some new python scripts which would be useful. 
@cclaver - note you can run "python testComponents.py" to check updates of the components vs. old versions - the old "intermediate files" are still present, but separated into an "IntermediateFiles" directory.
